### PR TITLE
Refactor Customer Class

### PR DESCRIPTION
**Pull request by NilsPetsch1999**
added 2 functions because of repeating lines (playaudio, updateSmiley) + simplified a few functions

### CODE REVIEW

File: src/main/java/com/example/decafe/Customer.java [hunk 1]
```diff
@@ -81,17 +77,8 @@ public ImageView getLabel() { //returns the label of the customer
...
-        Random random = new Random();
-        int number = random.nextInt(2);
-
-        switch (number) {
-            case 0 -> order = "cake";
-            case 1 -> order = "coffee";
-        }
-
-        return order;
+    public String getRandomOrder() {
```
Comment (Mu2menM) [inline src/main/java/com/example/decafe/Customer.java:82]: Can you please use here the CafeAssets for the strings : CafeAssets.CAKE.getName() and CafeAssets.COFFEE.getName()

File: src/main/java/com/example/decafe/Customer.java [hunk 2]
```diff
@@ -109,99 +96,53 @@ public static void setControllerTimer(Timer controllerTimer) { //sets the timer
 
     // Method used to create an Image Object
```
Comment (Mu2menM) [inline src/main/java/com/example/decafe/Customer.java]: Remove this function and make the customer class using the single responsibility method in ImageUtil : ImageUtil.getImageFromResources(filename)

File: src/main/java/com/example/decafe/Customer.java [hunk 3]
```diff
@@ -239,45 +180,32 @@ public void run() {
...
+
             }
         };
+        sixtySecondsTimer.schedule(timerTask, 0, 1000);
+    }
 
-        this.sixtySecondsTimer.schedule(timerTask, 0, 1000);//it should call this methode every second
-
+    private void updateSmiley(int seconds) {
+        try {
```
Comment (Mu2menM) [inline src/main/java/com/example/decafe/Customer.java:197]: Add the used image names to ImageAssets and call them from the ImageAssets enum : ImageAssets.SMILEY_GREEN.getName()

File: src/main/java/com/example/decafe/Customer.java [hunk 4]
```diff
@@ -81,17 +77,8 @@ public ImageView getLabel() { //returns the label of the customer
...
-        int number = random.nextInt(2);
-
-        switch (number) {
-            case 0 -> order = "cake";
-            case 1 -> order = "coffee";
-        }
-
-        return order;
+    public String getRandomOrder() {
+        return (order = new Random().nextBoolean() ? "cake" : "coffee");
```
Comment (albiosz) [inline src/main/java/com/example/decafe/Customer.java]: Why not just?:
```java
return new Random().nextBoolean() ? "cake" : "coffee"
```

File: src/main/java/com/example/decafe/Customer.java [hunk 5]
```diff
@@ -109,99 +96,53 @@ public static void setControllerTimer(Timer controllerTimer) { //sets the timer
...
-        } else if (customerImages[5].equals(customer)) {
-            wantedImage = searchArray[5];
-        } else if (customerImages[6].equals(customer)) {
-            wantedImage = searchArray[6];
+        for (int i = 0; i < customerImages.length; i++) {
+            if (customerImages[i].equals(customer)) return searchArray[i];
         }
-
-        return wantedImage;
+        return new ImageView();
```
Comment (albiosz) [inline src/main/java/com/example/decafe/Customer.java:104]: Now it looks reasonable :+1:

File: src/main/java/com/example/decafe/Customer.java [hunk 6]
```diff
@@ -109,99 +96,53 @@ public static void setControllerTimer(Timer controllerTimer) { //sets the timer
...
-            File f = new File("");
-            String musicFile = f.getAbsolutePath() + File.separator + "src" + File.separator + "main" + File.separator + "resources" + File.separator + "com" + File.separator + "example" + File.separator + "decafe" + File.separator + "doorBell.mp3";
-            AudioClip doorBell = new AudioClip(new File(musicFile).toURI().toString());
-            //MediaPlayer doorBell = new MediaPlayer(sound);
-            doorBell.play();
-            customer.waitingTime(); //place customer in the waitingTime of  60 seconds
+        if (customersInCoffeeShop.size() < 3 && !freeChairs.isEmpty()) {
+            ImageView customerImage = getRandomPic();
+            customerImage.setVisible(true);
+            Customer customer = new Customer(customerImage, getLabel(customerImage), freeSeatChosen, getImage(customerImage, smileyImages), getImage(customerImage, coinImages));
```
Comment (albiosz) [inline src/main/java/com/example/decafe/Customer.java]: From my perspective putting the four lines into one made the code less readable. We also lost the information about what exactly the customer receives (order, smiley, coin).

The article is about Python, but I think the practice is universal:
https://www.linkedin.com/pulse/zen-python-sparse-better-than-dense-paritosh-sharma-ghimire-bez5e/

File: src/main/java/com/example/decafe/Customer.java [hunk 7]
```diff
@@ -239,45 +180,32 @@ public void run() {
...
             }
         };
+        sixtySecondsTimer.schedule(timerTask, 0, 1000);
+    }
 
-        this.sixtySecondsTimer.schedule(timerTask, 0, 1000);//it should call this methode every second
-
+    private void updateSmiley(int seconds) {
+        try {
+            String imageName = seconds == 59 ? "smileygreen.png" : seconds == 30 ? "smileyyellow.png" : "smileyred.png";
```
Comment (albiosz) [inline src/main/java/com/example/decafe/Customer.java]: From my perspective, nested ternary operators decrease readability of the code, make it more dense.

https://www.sonarsource.com/blog/stop-nesting-ternaries-javascript/

File: src/main/java/com/example/decafe/Customer.java [hunk 8]
```diff
@@ -239,45 +180,32 @@ public void run() {
...
+            smiley.setVisible(true);
+            smiley.setImage(createImage(imageName));
+            setMood(seconds == 59 ? CustomerMode.GREEN : seconds == 30 ? CustomerMode.YELLOW : CustomerMode.RED);
+        } catch (FileNotFoundException e) {
+            e.printStackTrace();
+        }
```
Comment (albiosz) [inline src/main/java/com/example/decafe/Customer.java]: It's nice that `smiley.setVisible(true)` is not repeated, but the numbers: 59, 30, 15 are now, so I do not think it's a big improvement, is it?

File: src/main/java/com/example/decafe/Customer.java [hunk 9]
```diff
@@ -337,29 +265,20 @@ public static void noMoneySpent(Customer customer) throws FileNotFoundException
...
-            String musicFile = f.getAbsolutePath() + File.separator + "src" + File.separator + "main" + File.separator + "resources" + File.separator + "com" + File.separator + "example" + File.separator + "decafe" + File.separator + "wrongChoice.mp3";
-            AudioClip wrongOrder = new AudioClip(new File(musicFile).toURI().toString());
-            //MediaPlayer collectMoney = new MediaPlayer(sound);
-            wrongOrder.play();
-            this.coinImage.setImage(this.createImage("coin.png")); // set coin Image to empty plate
-            this.coinImage.setOnMouseClicked(event1 -> { // set click event to this
+        customersInCoffeeShop.removeIf(customer -> customer.getImage().equals(customerImage));
+        coinImage.setVisible(true);
+        coinImage.setDisable(false);
+        playAudio(leftUnhappy ? "wrongChoice.mp3" : "rightChoice.mp3");
```
Comment (albiosz) [inline src/main/java/com/example/decafe/Customer.java:275]: That's a nice improvement!

**Review by albiosz — CHANGES_REQUESTED** [review]
Some suggestions that would improve readability from my perspective. I really like the topic of clean code, so I'm always happy to discuss about it ;)

**Comment by NilsPetsch1999** [comment]
Made a few changes after looking through your recommendations :) pls review again @albiosz 
integrated CafeAssets, ImageAssets and ImageUtil to Customer Class. tried to implement readability recommendations

File: src/main/java/com/example/decafe/Customer.java (hunk 8 above)
Comment (albiosz) [reply src/main/java/com/example/decafe/Customer.java]: I had no idea, how to refactor it better, but I got a nice suggestion from ChatGPT ;)
```java
public void waitingTime() {
    TimerTask timerTask = new TimerTask() {
        int seconds = 60;

        @Override
        public void run() {
            seconds--;
            if (seconds == 0) {
                startTimerLeave(Customer.this);
            } else {
                updateSmiley(seconds);
            }
        }
    };
    sixtySecondsTimer.schedule(timerTask, 0, 1000);
}

private void updateSmiley(int seconds) {
    try {
        String imageName = getImageName(seconds);
        CustomerMode mood = getCustomerMode(seconds);
        if (imageName != null && mood != null) {
            setMood(mood);
            smiley.setVisible(true);
            smiley.setImage(ImageUtil.getImageFromResources(imageName));
        }
    } catch (ImageNotFoundException e) {
        e.printStackTrace();
    }
}

private String getImageName(int seconds) {
    switch (seconds) {
        case 59:
            return ImageAssets.CUSTOMER_SMILEY_GREEN.getImage();
        case 30:
            return ImageAssets.CUSTOMER_SMILEY_YELLOW.getImage();
        case 15:
            return ImageAssets.CUSTOMER_SMILEY_RED.getImage();
        default:
            return null;
    }
}

private CustomerMode getCustomerMode(int seconds) {
    switch (seconds) {
        case 59:
            return CustomerMode.GREEN;
        case 30:
            return CustomerMode.YELLOW;
        case 15:
            return CustomerMode.RED;
        default:
            return null;
    }
}
```

**Review by albiosz — APPROVED** [review]
Just added a suggestion, how `waitingTime()` and `updateSmiley()` could be refactored, but it does not need to be applied. Apart from it, it can be merged :+1:

**Pull request merged**